### PR TITLE
Add native Linux syscall interface

### PR DESCRIPTION
**Pull request by lbguilherme**
This is the first piece from #10768, adding a `Crystal::System::Syscall` module. It would be capable to provide any system call, but only the ones required by io_uring are defined.

Here's a small example showing how to use it:

```cr
require "crystal/system/unix/syscall/*"

module Crystal::System::Syscall
  def_syscall write, Int32, fd : Int32, buf : UInt8*, count : LibC::SizeT
end

data = "Hello!\n"
Crystal::System::Syscall.write(1, data.to_unsafe, LibC::SizeT.new(data.size))
```

### CODE REVIEW

File: src/crystal/system/unix/syscall/linux.cr
Comment (yxhuvud) [inline src/crystal/system/unix/syscall/linux.cr:288]: This one (and msg_header, later when the method that needs that is implemented), should probably be in `LibC`.

File: src/crystal/system/unix/syscall/arm-linux.cr
Comment (kubo) [inline src/crystal/system/unix/syscall/arm-linux.cr:407]: According to "Architecture calling conventions" in [this man page](https://man7.org/linux/man-pages/man2/syscall.2.html):

Arch/ABI | Instruction | System Call # | Ret Val | arg1 | arg2 | arg3 | arg4 | arg5 | arg6
---|---|---|---|---|---|---|---|---|---
arm/EABI  |  `swi 0x0` | r7 | r0 | r0  | r1  | r2 |  r3  | r4 |  r5
arm64  |    `svc #0`    |   w8   |   x0 | x0  |  x1  |  x2  |  x3   | x4  |  x5

**Comment by straight-shoota** [comment]
I think all the parts regarding io_uring should not be in this PR. Let's keep it focused on syscalls in general and not targeted at a specific use case. That probably affects everything in `linux.cr`.

Instead, I would like to see some specs. Adding some simple syscalls and check that they work as expected shouldn't be too much effort.

I'm also wondering if the API shouldn't be public instead of hidden in `Crystal::System`. There are good use cases for defining syscalls in user code. Your io_uring shard is an example of that.

**Comment by yxhuvud** [comment]
@straight-shoota my shard depends on liburing and thus doesn't need access to syscalls on the crystal side. 

Don't we have some existing raw syscalls emitted btw? I'm thinking related to changing fibers but perhaps we have more examples? Unifying that a little could be a nice followup pr from someone once these are merged.

**Comment by straight-shoota** [comment]
The shard could reasonably use syscalls instead of liburing. And other shards could use syscalls for other purposes.

Good point about porting existing syscalls. I believe the only one in use is `getrandom`.

**Comment by straight-shoota** [comment]
I would suggest not to expose any actual syscall definitions in the public API. At least not until we're certain that's what we want to do. I feel like we still need to form an opinion on that. Removing methods after they've been exposed in the public API is hard, but if we keep them internal for now, it's always easy to make them public.

For now, I think just defining `Syscall.def_syscall` should be sufficient, with only internal definitions for spec and `Crystal::System::Random`.

The actual syscall definitions could either be nodoc in `Syscall`, or even just locally scoped (e.g. `private Syscall.def_syscall write(...` in the spec file).

I'd like to hear other's opinions on this, though. But I think it's the safer way for now, until we've figured out what we want syscall definitions to look like.

**Comment by HertzDevil** [comment]
We could always put `@[Experimental]` on the syscall API, and I believe that annotation is added for precisely this kind of situation now that we've reached 1.0.

**Comment by lbguilherme** [comment]
New set of changes:

1. Marked as `@[Experimental]`.
2. No syscalls are defined in a public way. They are either private to System::Random or to the spec.

File: spec/std/syscall_spec.cr
Comment (straight-shoota) [inline spec/std/syscall_spec.cr:6]: This private namespace shouldn't be necessary (same for random). `Syscall.def_syscall` should just be a public macro and callable from anywhere. Then the following defs can be `private Syscall.def_syscall xxx` in the file scope.

Comment (lbguilherme) [reply spec/std/syscall_spec.cr]: Done.

File: src/crystal/system/unix/getrandom.cr
Comment (straight-shoota) [inline src/crystal/system/unix/getrandom.cr:6]: I think the method should still be namespaced to `Crystal::System::Random`, not in the top-level namespace.

File: src/syscall/aarch64-linux.cr
Comment (straight-shoota) [inline src/syscall/aarch64-linux.cr:5]: This enum is internal, so it should be `nodoc`, right?

File: src/crystal/system/unix/getrandom.cr
Comment (lbguilherme) [reply src/crystal/system/unix/getrandom.cr:6]: This will conflict with `Crystal::System::Random.getrandom`, which already exists. One possible solution would be to make `def_syscall` always generate a `sys_` prefix and then rename the existing `sys_getrandom` to `sys_getrandom_wrap`. What do you think?

File: src/syscall/aarch64-linux.cr
Comment (lbguilherme) [reply src/syscall/aarch64-linux.cr:6]: Fixed.

**Review by straight-shoota — APPROVED** [review]

File: src/crystal/system/unix/getrandom.cr
Comment (straight-shoota) [reply src/crystal/system/unix/getrandom.cr:6]: Depends on how we want to proceed with syscalls in the stdlib API. 
If we end up collecting all the methods in `Syscall` as you initially proposed, there's not much reason to think about naming conflicts. So let's wait and see.

I think it's best to leave like that for now.

Comment (asterite) [reply src/crystal/system/unix/getrandom.cr:6]: Why not bind to Libc?

Comment (lbguilherme) [reply src/crystal/system/unix/getrandom.cr:6]: It would require glibc 2.25. Is that OK?

Comment (asterite) [reply src/crystal/system/unix/getrandom.cr:6]: Not sure. It is something super new?

Comment (asterite) [reply src/crystal/system/unix/getrandom.cr:6]: That said, it's fine for now

Comment (yxhuvud) [reply src/crystal/system/unix/getrandom.cr:6]: @asterite This particular change is just cleanup changing an existing syscall that is already used in the current code to instead use a little more structured way to define syscalls. I don't know if this particular one exist in LibC yet.

However, the relevant syscalls driving the feature is the ones related to io_uring. Those are from Linux 5.3 or something like that, and LibC has not included them yet. There is a more user friendly C library for uring that the kernel maintainers also maintain at https://github.com/axboe/liburing , but that can also not be expected to exist as a distro-packaged library.

**Comment by mgomes** [comment]
What's remaining to get this merged in? Anything we can do to help?

**Comment by straight-shoota** [comment]
This is waiting for review by another Core Team member.

**Comment by asterite** [comment]
Can iouring be done without this? I don't understand why we need to leak all of this. It seems we already have `getrandom` in place by using `LibC.syscall`?

I just wonder whether all this complexity is worth it for just having `getrandom` called in a different way 🤔 

Also, supporting direct syscalls in the interpreter will be a lot of work.

**Comment by yxhuvud** [comment]
> Can iouring be done without this?

Assuming `this` refer to supporting raw syscalls, then yes. That would however force us to instead link against [liburing](https://github.com/axboe/liburing), which in turn does the raw syscalls. It is a different (but nicer and simpler) API though and as such it would force the implementation in the later parts to be rewritten more or less from scratch.

**Comment by asterite** [comment]
What are the syscalls that we need to do? Because, looking at #10768, it looks like it's only using `getrandom`, which we already have in place without having to use this PR.

I guess I'm not seeing the full picture here... are there other syscalls that #10768 will eventually use? If that's the case, can this PR be delayed until that's needed?

**Comment by asterite** [comment]
Oh, nevermind, I see there `io_uring_setup`...

File: src/crystal/system/unix/getrandom.cr
Comment (asterite) [reply src/crystal/system/unix/getrandom.cr:6]: I agree with @straight-shoota , it feels a bit weird that the syscall is defined here at the top-level. Privately, but at the top-level.

Why not all syscalls are defined directly inside the `Syscall` namespace? Then you can do `Syscall.getrandom` from anywhere, and it's very clear that `getrandom` is a syscall...

**Comment by yxhuvud** [comment]
@asterite `.._setup` is for initializing the ring. There is also `io_uring_enter`, which is used for submitting and fetching events.

**Comment by straight-shoota** [comment]
Using `liburing` obviously has the benefit of a more polished API and less for us to do ourselves. But the syscall-based implementation is already done, so switching would have a higher cost. Not sure about the impact of long term maintenance. But using syscall directly sounds like a better solution. It also avoids an additional binary dependency.

I suppose a follow-up question would be if we need this native syscall interface for `io_uring` or could it work with `LibC.syscall`? At least for a start.

However, I think it makes sense to have a native syscall interface in Crystal eventually, so why not do it now.

**Comment by asterite** [comment]
I agree with everything you said. Thank you! ❤️ 

That said, I can't approve this until the generated syscall methods leaves under a namespace instead of the top-level.

**Comment by lbguilherme** [comment]
@asterite I made a change forcing syscalls to be defined inside the `Syscall` module. Is that what you meant?

**Review by asterite — APPROVED** [review]
Looks great, thank you! ❤️

File: src/syscall.cr
Comment (HertzDevil) [inline src/syscall.cr:17]: The docs must now be placed at `Syscall` itself because docs for private macros are not shown.

Comment (lbguilherme) [reply src/syscall.cr:23]: Fixed, thanks!

File: src/crystal/system/unix/getrandom.cr
Comment (beta-ziliani) [reply src/crystal/system/unix/getrandom.cr]: But there's a point in having syscalls within the `Crystal::System` namespace, to make sure they're not consider part of the public API. So I think it would be best to go with `Crystal::System::Syscall`.

File: src/syscall.cr
Comment (beta-ziliani) [reply src/syscall.cr:23]: Why making it `private`? This is fixing the module in which syscall are placed, but we don't do that for other bindings (we just use a convention instead). Like with LibC bindings, for instance, we except shards to define their own syscalls in their own namespace.

**Review by beta-ziliani — COMMENTED** [review]
I left two comments (the first one fell out of the review: https://github.com/crystal-lang/crystal/pull/10777#discussion_r745627319, I clicked the wrong button 🤦 )

**Comment by lbguilherme** [comment]
It is unclear if the `def_syscall` macro should be part of the public API (marked as experimental for now) or if it should be private, inside the `Crystal` namespace.

I have seen some conflicting opinions here. So, a few options about how to move forward, in no particular order:

1- `Syscall` is a public top-level module from the stdlib. `def_syscall` is a public macro and users and encouraged to define their own modules and define their syscalls inside it using `def_syscall`.

2- `Crystal::System::Syscall` is an internal module and the syscalls that the stdlib uses are already defined there.

3- We stick with `LibC.syscall` for now (changing the io_uring pr to use it) and don't introduce anything to the stdlib.

4- A shard is made with definitions for all Linux system calls and anyone can use it. Then maybe at some point, the stdlib decides to use it as a subtree in a private namespace for its internal implementations.

5- io_uring is remade using `liburing` and we figure out how to solve the distribution problem of this new dependency.

What way should we go?

**Comment by beta-ziliani** [comment]
Thanks for the nice summary of options. My vote goes for 2.

**Comment by straight-shoota** [comment]
@asterite I totally agree. But I assume that's what the `Experimental` annotation is for. It allows to expose a public API so people can start using it (with some reservations) and we can learn how that works best. But we don't commit to a strong backwards compatibility for it and reserve the right to immediate changes.
So IMO `def_syscall` should be public with `Experimental` state (the entire annotation is still an experiment itself, but we won't get anywhere if we don't use it - just to repeat the argument).

The methods defined by `def_syscall` however I would prefer to put in the private `Crystal::System` namespace. `Crystal::System::Syscall` sounds like a good idea. So that's option 2. 👍 

This doesn't necessarily exclude that we might eventually move to option 1 at some point. But I'd rather start some experimenting phase (hence make `def_syscall` public) before deciding if we should commit to a global namespace. The syscalls themselves are global, but the bindings might not be, similar to lib bindings.

**Comment by lbguilherme** [comment]
I pushed a new attempt:

- There is an `@[Experimental]` public module to allow defining syscalls. The user is encouraged to define their syscalls in a new module to avoid namespace pollution.
- All syscalls defined by the stdlib itself are in `Crystal::System::Syscall`.

File: src/syscall.cr
Comment (straight-shoota) [inline src/syscall.cr:22]: The source code blocks need to be indented by four spaces, or delimited by triple backticks. I prefer backticks because that's the more common style in stdlib.

**Review by straight-shoota — APPROVED** [review]

File: src/syscall.cr
Comment (lbguilherme) [reply src/syscall.cr]: Fixed.

**Review by beta-ziliani — APPROVED** [review]
Thanks a lot @lbguilherme ! 🙇

**Review by asterite — APPROVED** [review]

**Pull request merged**
